### PR TITLE
JSDoc support

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Adds support for adding type information in JavaScript files using jsdoc tags

### CODE REVIEW

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -227,6 +228,19 @@ namespace ts {
...
                     return node.flags & NodeFlags.Default ? "default" : undefined;
+                case SyntaxKind.JSDocFunctionType:
+                    return isJSDocConstructSignature(node) ? "__new" : "__call";
+                case SyntaxKind.Parameter:
+                    // Parameters with names are handled at the top of this function.  Parameters
+                    // without names can only come from JSDocFunctionTypes.
+                    Debug.assert(node.parent.kind === SyntaxKind.JSDocFunctionType);
+                    let functionType = <JSDocFunctionType>node.parent;
+                    let index = indexOf(functionType.parameters, node);
+                    return "p" + index;
```
Comment (DanielRosenwasser) [inline src/compiler/binder.ts]: Will we ever manifest this name to the user?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -1198,6 +1199,21 @@ namespace ts {
                             ));
                         }
                     }
+
+                    // CommonJS 'module.exports = expr' assignments
+                    const commonJsModuleExports = symbol.exports["__jsExports"];
+                    if (commonJsModuleExports) {
+                        for (let i = 0; i < commonJsModuleExports.declarations.length; i++) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Extract out declarations.

File: src/compiler/parser.ts [hunk 3]
```diff
@@ -611,44 +611,26 @@ namespace ts {
...
-        function addJSDocComment(node: Node) {
-            const comments = getLeadingCommentRangesOfNode(node, sourceFile);
-            if (comments) {
-                for (const comment of comments) {
-                    const jsDocComment = JSDocParser.parseJSDocComment(node, comment.pos, comment.end - comment.pos);
-                    if (jsDocComment) {
-                        node.jsDocComment = jsDocComment;
+        function addJSDocComment<T extends Node>(node: T): T {
+            if (contextFlags & ParserContextFlags.JavaScriptFile) {
+                speculationHelper(() => {
```
Comment (ahejlsberg) [inline src/compiler/parser.ts]: Call `lookAhead`.

Comment (ahejlsberg) [inline src/compiler/parser.ts]: Actually, something is fishy here, we shouldn't need the `speculationHelper` call at all.

Comment (RyanCavanaugh) [inline src/compiler/parser.ts]: Removed

File: src/compiler/binder.ts (hunk 1 above)
Comment (RyanCavanaugh) [inline src/compiler/binder.ts]: Yes

File: src/compiler/parser.ts [hunk 4]
```diff
@@ -611,44 +611,24 @@ namespace ts {
...
-                    if (jsDocComment) {
-                        node.jsDocComment = jsDocComment;
+        function addJSDocComment<T extends Node>(node: T): T {
+            if (contextFlags & ParserContextFlags.JavaScriptFile) {
+                const comments = getLeadingCommentRangesOfNode(node, sourceFile);
+                if (comments) {
+                    for (const comment of comments) {
+                        const jsDocComment = JSDocParser.parseJSDocComment(node, comment.pos, comment.end - comment.pos);
+                        if (jsDocComment) {
+                            node.jsDocComment = jsDocComment;
```
Comment (ahejlsberg) [inline src/compiler/parser.ts]: So if there's more than one JSDoc comment, the last one wins?

Comment (RyanCavanaugh) [inline src/compiler/parser.ts]: Yes, that's intentional.

It's likely to see things like this
```ts
/** This is a comment for a commented-out function
  * @param foo
  */
/* function removedTemp(foo) {
 ...
}
*/
/** This is the real JSDoc Comment
  * @param foo
  */
function realFunc(foo) { ... }
```

or this

```ts
/** This is just a general comment that applies to the whole file
  * @summary You can use this file for neat things
  */
/** And here's a function
  * @summary Don't call this
  */
function fn() { ... }
```

File: src/compiler/parser.ts [hunk 5]
```diff
@@ -611,44 +611,24 @@ namespace ts {
                 fixupParentReferences(sourceFile);
```
Comment (ahejlsberg) [inline src/compiler/parser.ts]: There are several calls to 'fixupParentReferences` from within the JSDoc parser, but only the above call is necessary because JSDoc comments are no longer parsed after the main pass.

File: src/compiler/parser.ts [hunk 6]
```diff
@@ -5580,23 +5561,19 @@ namespace ts {
...
+                const jsDocTypeExpression = parseJSDocTypeExpression();
                 const diagnostics = parseDiagnostics;
                 clearState();
 
                 return jsDocTypeExpression ? { jsDocTypeExpression, diagnostics } : undefined;
             }
 
-            // Parses out a JSDoc type expression.  The starting position should be right at the open
-            // curly in the type expression.  Returns 'undefined' if it encounters any errors while parsing.
+            // Parses out a JSDoc type expression. Returns 'undefined' if it encounters any errors while parsing.
```
Comment (ahejlsberg) [inline src/compiler/parser.ts]: I don't think this ever returns undefined.

**Comment by ahejlsberg** [comment]
The `parseJSDocOptionalType` function in parser.ts is never called.

File: src/compiler/binder.ts [hunk 7]
```diff
@@ -1241,6 +1265,8 @@ namespace ts {
 
                 case SyntaxKind.CallExpression:
                     if (isInJavaScriptFile(node)) {
+                        // We're only inspecting call expressions to detect CommonJS modules, so we can skip
+                        // this check if we've already seen the module indicator
```
Comment (DanielRosenwasser) [inline src/compiler/binder.ts]: I'm not seeing where you skip this check.

File: src/compiler/binder.ts [hunk 8]
```diff
@@ -1411,6 +1436,7 @@ namespace ts {
             // this check if we've already seen the module indicator
             if (!file.commonJsModuleIndicator && isRequireCall(node)) {
                 setCommonJsModuleIndicator(node);
+                declareSymbol(file.symbol.exports, file.symbol, node, SymbolFlags.None, SymbolFlags.None);
```
Comment (billti) [inline src/compiler/binder.ts]: I'm not clear why this change is in this JsDoc work. Can you explain?

File: src/compiler/checker.ts [hunk 9]
```diff
@@ -8350,10 +8650,7 @@ namespace ts {
         }
 
         function getJsxGlobalElementClassType(): Type {
-            if (!jsxElementClassType) {
-                jsxElementClassType = getExportedTypeFromNamespace(JsxNames.JSX, JsxNames.ElementClass);
-            }
-            return jsxElementClassType;
+            return getExportedTypeFromNamespace(JsxNames.JSX, JsxNames.ElementClass);
```
Comment (billti) [inline src/compiler/checker.ts]: Are the above supposed to be in this change? (See removal of var on line 167 also).

File: src/compiler/checker.ts [hunk 10]
```diff
@@ -1218,6 +1219,22 @@ namespace ts {
...
+                            const properties = getPropertiesOfType(checkExpression((<BinaryExpression>declarations[i]).right));
+                            if (i === 0) {
+                                symbols = createSymbolTable(properties);
+                            }
+                            else {
+                                mergeSymbolTable(symbols, createSymbolTable(properties));
+                            }
+                        }
+                    }
+
```
Comment (billti) [inline src/compiler/checker.ts]: I assume a bad merge? Doesn't look JsDoc related.

File: src/compiler/checker.ts [hunk 11]
```diff
@@ -546,7 +547,7 @@ namespace ts {
                             // - Type parameters of a function are in scope in the entire function declaration, including the parameter
                             //   list and return type. However, local types are only in scope in the function body.
                             // - parameters are only in the scope of function body
-                            if (meaning & result.flags & SymbolFlags.Type) {
+                            if (meaning & result.flags & SymbolFlags.Type && lastLocation.kind !== SyntaxKind.JSDocComment) {
```
Comment (billti) [inline src/compiler/checker.ts]: Can you add a brief comment as to why this check is needed here? I'm having trouble groking where this might be an issue based on how this while loop walks up the parents.

File: src/compiler/checker.ts [hunk 12]
```diff
@@ -1187,7 +1188,7 @@ namespace ts {
                     return;
                 }
                 visitedSymbols.push(symbol);
-                const symbols = cloneSymbolTable(symbol.exports);
+                let symbols = cloneSymbolTable(symbol.exports);
```
Comment (billti) [inline src/compiler/checker.ts]: Unrelated to JsDoc?

File: src/compiler/checker.ts [hunk 13]
```diff
@@ -3869,6 +3966,10 @@ namespace ts {
                 const signatureDeclaration = <SignatureDeclaration>node.parent;
                 const signature = getSignatureFromDeclaration(signatureDeclaration);
                 const parameterIndex = ts.indexOf(signatureDeclaration.parameters, node);
+                if (skipSignatureCheck) {
```
Comment (billti) [inline src/compiler/checker.ts]: Why is this check this far down and specific to nodes with initializers? (I'd expect it to be at least 3 lines up to avoid that work regardless).

File: src/compiler/checker.ts [hunk 14]
```diff
@@ -4196,9 +4317,10 @@ namespace ts {
         }
 
         // Get type from reference to class or interface
-        function getTypeFromClassOrInterfaceReference(node: TypeReferenceNode | ExpressionWithTypeArguments, symbol: Symbol): Type {
+        function getTypeFromClassOrInterfaceReference(node: TypeReferenceNode | ExpressionWithTypeArguments | JSDocTypeReference, symbol: Symbol): Type {
```
Comment (billti) [inline src/compiler/checker.ts]: It might be more readable just to create a type alias for "TypeReferenceNode | ExpressionWithTypeArguments | JSDocTypeReference". It's used repeatedly (7 times) in the following code.

File: src/compiler/checker.ts [hunk 15]
```diff
@@ -4640,9 +4892,37 @@ namespace ts {
...
+                    return getTypeFromJSDocVariadicType(<JSDocVariadicType>node);
+                case SyntaxKind.JSDocConstructorType:
+                    return getTypeFromTypeNode((<JSDocConstructorType>node).type);
+                case SyntaxKind.JSDocRecordType:
+                    return getTypeFromJSDocRecordType(<JSDocRecordType>node);
+                case SyntaxKind.JSDocThisType:
+                    return getTypeFromTypeNode((<JSDocThisType>node).type);
             }
+
+            return unknownType;
```
Comment (billti) [inline src/compiler/checker.ts]: I'd rather keep the `default:` on the switch statement still return the unknownType, than fall out of the switch statement with no default and return it.

File: src/compiler/checker.ts [hunk 16]
```diff
@@ -9155,34 +9452,34 @@ namespace ts {
           */
         function getEffectiveDecoratorFirstArgumentType(node: Node): Type {
```
Comment (billti) [inline src/compiler/checker.ts]: Not sure where all the below changes are coming from, but not JsDoc related.

File: src/compiler/checker.ts [hunk 17]
```diff
@@ -9982,13 +10287,14 @@ namespace ts {
         }
 
         function getReturnTypeFromBody(func: FunctionLikeDeclaration, contextualMapper?: TypeMapper): Type {
-            const contextualSignature = getContextualSignatureForFunctionLikeDeclaration(func);
             if (!func.body) {
                 return unknownType;
             }
 
+            const contextualSignature = getContextualSignatureForFunctionLikeDeclaration(func);
```
Comment (billti) [inline src/compiler/checker.ts]: Unrelated to JsDoc?

File: src/compiler/types.ts [hunk 18]
```diff
@@ -389,6 +389,8 @@ namespace ts {
         ContainsThis =      1 << 18,  // Interface contains references to "this"
         HasImplicitReturn =     1 << 19,  // If function implicitly returns on one of codepaths (initialized by binding)
         HasExplicitReturn =     1 << 20,  // If function has explicit reachable return on one of codepaths (initialized by binding)
+        InferredClass =         1 << 21,  // An ES6 class
```
Comment (billti) [inline src/compiler/types.ts]: JsDoc related?

File: src/compiler/parser.ts (hunk 5 above)
Comment (RyanCavanaugh) [inline src/compiler/parser.ts]: One of them was needed and one wasn't -- removed the unneeded one

File: src/compiler/parser.ts (hunk 6 above)
Comment (RyanCavanaugh) [inline src/compiler/parser.ts]: Removed

File: src/compiler/checker.ts (hunk 13 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Good point

File: src/compiler/checker.ts (hunk 14 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Nathan and I tried for a while and couldn't come up with a good name that wasn't taken. Open to suggestions.

File: src/compiler/checker.ts (hunk 15 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Sure

File: src/compiler/checker.ts (hunk 14 above)
Comment (weswigham) [inline src/compiler/checker.ts]: `TypeReferenceContainingNode`?

Comment (sandersn) [inline src/compiler/checker.ts]: Sounds pretty good to me.

File: src/compiler/scanner.ts [hunk 19]
```diff
@@ -1649,6 +1656,69 @@ namespace ts {
...
+                        return pos += 1, token = SyntaxKind.EqualsToken;
+                    }
+                    else if (ch === CharacterCodes.comma) {
+                        return pos += 1, token = SyntaxKind.CommaToken;
+                    }
+                    else if (isWhiteSpace(ch)) {
+                        // Keep going
+                    }
+                    else {
+                        identifierStarted = true;
```
Comment (ahejlsberg) [inline src/compiler/scanner.ts]: Hmm, so an identifier can start with *anything* but the characters checked for above, but then after that must contain proper identifier characters. That seems a bit odd.

File: src/compiler/parser.ts [hunk 20]
```diff
@@ -6110,95 +6103,88 @@ namespace ts {
...
-                        }
-
-                        break;
-                    }
+                function nextJSDocToken(): SyntaxKind {
+                    return token = scanner.scanJSDocToken();
+                }
 
-                    if (startPos === pos) {
+                function scanJsDocIdentifier(): Identifier {
```
Comment (ahejlsberg) [inline src/compiler/parser.ts]: Should be `parseJSDocIdentifier`. The scanning is handled by the scanner.

Comment (ahejlsberg) [inline src/compiler/parser.ts]: Also, it should call `nextJSDocToken()`, similar to what `parseIdentifier` in the main parser does.

File: src/compiler/scanner.ts [hunk 21]
```diff
@@ -1664,6 +1671,72 @@ namespace ts {
...
+                        return pos += 1, token = SyntaxKind.CloseBracketToken;
+                    }
+                    else if (ch === CharacterCodes.equals) {
+                        return pos += 1, token = SyntaxKind.EqualsToken;
+                    }
+                    else if (ch === CharacterCodes.comma) {
+                        return pos += 1, token = SyntaxKind.CommaToken;
+                    }
+                    else if (isWhiteSpace(ch)) {
+                        // Keep going
```
Comment (billti) [inline src/compiler/scanner.ts]: This should never happen though right? As you've eaten the preceding white space, and there's nothing in here where you would consume white space inside a token.

File: src/compiler/scanner.ts [hunk 22]
```diff
@@ -1686,6 +1759,33 @@ namespace ts {
...
+
+            end = saveEnd;
+            pos = savePos;
+            startPos = saveStartPos;
+            tokenPos = saveTokenPos;
+            token = saveToken;
+            precedingLineBreak = savePrecedingLineBreak;
+            tokenValue = saveTokenValue;
+            hasExtendedUnicodeEscape = saveHasExtendedUnicodeEscape;
+            tokenIsUnterminated = saveTokenIsUnterminated;
```
Comment (billti) [inline src/compiler/scanner.ts]: It seems a little fragile to use  a method (`setText` - which calls `setTextPos`) to modify the state prior to callback execution, then manually restore each field individually after. If anyone added further state changes in `setText` (or `setTextPos`), they'd need to know to come and manually undo those same changes here. Can this set/restore be made more symmetrical/colocated somehow?

File: src/compiler/binder.ts [hunk 23]
```diff
@@ -873,7 +888,7 @@ namespace ts {
             }
         }
 
-        function bindFunctionOrConstructorType(node: SignatureDeclaration) {
+        function bindFunctionOrConstructorTypeOrJSDocFunctionType(node: SignatureDeclaration): void {
```
Comment (ahejlsberg) [inline src/compiler/binder.ts]: I'd prefer to just keep the existing name.

File: src/compiler/checker.ts [hunk 24]
```diff
@@ -3860,12 +3923,33 @@ namespace ts {
             return result;
         }
 
-        function isOptionalParameter(node: ParameterDeclaration) {
+        function isOptionalParameter(node: ParameterDeclaration, skipSignatureCheck?: boolean) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: The `skipSignatureCheck` parameter never appears to be used.

File: src/compiler/scanner.ts (hunk 22 above)
Comment (RyanCavanaugh) [inline src/compiler/scanner.ts]: This is the same pattern we use elsewhere (e.g. in `speculationHelper`). I don't see any obvious improvement.

File: src/server/server.ts [hunk 25]
```diff
@@ -7,12 +7,58 @@ namespace ts.server {
...
+    // Need to write directly to stdout, else rl.write also causes an input "line" event
+    // See https://github.com/joyent/node/issues/4243
+    let writeHost = (data: string) => process.stdout.write(data);
+
+    // Stubs for I/O
+    const onInput = (input: string) => { return; };
+    const onClose = () => { return; };
+
+    // Use a socket for comms if defined
+    const tss_debug: string = process.env["TSS_DEBUG"];
```
Comment (billti) [inline src/server/server.ts]: All these edits to server.ts look like changes I made in my branch to help with debugging Sublime but hadn't pushed to 'master' (and are part of bigger changes on both Sublime and tsserver to use TCP over stdin/stdout if configured to do so via `TSS_DEBUG`). Any idea how the changes in this file got in this code review?

File: src/compiler/checker.ts [hunk 26]
```diff
@@ -3838,6 +3886,21 @@ namespace ts {
             return getIndexTypeOfStructuredType(getApparentType(type), kind);
         }
 
+        function getTypeParametersFromSignatureDeclaration(declaration: SignatureDeclaration): TypeParameter[] {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: I think this should be `getTypeParametersFromJSDocTemplate` or something such.

File: src/compiler/checker.ts [hunk 27]
```diff
@@ -3838,6 +3886,21 @@ namespace ts {
...
 
+        function getTypeParametersFromSignatureDeclaration(declaration: SignatureDeclaration): TypeParameter[] {
+            if (declaration.parserContextFlags & ParserContextFlags.JavaScriptFile) {
+                const templateTag = getJSDocTemplateTag(declaration);
+                if (templateTag) {
+                    return getTypeParametersFromDeclaration(templateTag.typeParameters);
+                }
+            }
+
+            if (declaration.typeParameters) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: This test will always be false (given the one place the function is called), so the `if` should be removed.

File: src/compiler/checker.ts [hunk 28]
```diff
@@ -3862,6 +3925,23 @@ namespace ts {
         }
 
         function isOptionalParameter(node: ParameterDeclaration) {
+            if (node.parserContextFlags & ParserContextFlags.JavaScriptFile) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: I never liked the extra `parserContextFlags` property that Cyrus introduced. Those flags should just be in the regular `flags` property where there is ample room. The problem is that most nodes don't have the property, so this will be a cache miss most of the time when you're not a JavaScript file. We really should just consolidate the two.

File: src/compiler/checker.ts [hunk 29]
```diff
@@ -4249,18 +4349,90 @@ namespace ts {
...
+
+            if (!typeReferenceName) {
+                return unknownSymbol;
+            }
+
+            let symbol = resolveEntityName(typeReferenceName, SymbolFlags.Type);
+            if (!symbol && node.kind === SyntaxKind.JSDocTypeReference) {
+                // If the reference didn't resolve to a type, try seeing if results to a 
+                // value.  If it does, get the type of that value.
+                symbol = resolveEntityName(typeReferenceName, SymbolFlags.Value);
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: I think I've commented before that this seems rather ad hoc. Are we sure that's the behavior we want?

File: src/compiler/checker.ts [hunk 30]
```diff
@@ -4249,18 +4349,90 @@ namespace ts {
...
-                    isSupportedExpressionWithTypeArguments(<ExpressionWithTypeArguments>node) ? (<ExpressionWithTypeArguments>node).expression :
-                        undefined;
-                const symbol = typeNameOrExpression && resolveEntityName(typeNameOrExpression, SymbolFlags.Type) || unknownSymbol;
-                const type = symbol === unknownSymbol ? unknownType :
-                    symbol.flags & (SymbolFlags.Class | SymbolFlags.Interface) ? getTypeFromClassOrInterfaceReference(node, symbol) :
-                        symbol.flags & SymbolFlags.TypeAlias ? getTypeFromTypeAliasReference(node, symbol) :
-                            getTypeFromNonGenericTypeReference(node, symbol);
+                let symbol: Symbol;
+                let type: Type;
+                if (node.kind === SyntaxKind.JSDocTypeReference) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: It seems this branch of the `if` is strictly a superset of the other branch. Could we just remove the test and the `else` branch.

File: src/compiler/checker.ts [hunk 31]
```diff
@@ -7063,6 +7269,13 @@ namespace ts {
                 return container.flags & NodeFlags.Static ? getTypeOfSymbol(symbol) : (<InterfaceType>getDeclaredTypeOfSymbol(symbol)).thisType;
             }
 
+            if (container.parserContextFlags & ParserContextFlags.JavaScriptFile) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: This `if` and the following `if` should both be in a single `if (isInJavaScriptFile(node))`.

**Comment by ahejlsberg** [comment]
Made it all the way through. I think we're good once my comments are addressed.

File: src/compiler/checker.ts (hunk 28 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 27 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 26 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/server/server.ts (hunk 25 above)
Comment (RyanCavanaugh) [inline src/server/server.ts]: Undone.

File: src/compiler/checker.ts (hunk 29 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: I'm not sure what the motivation behind this was. I'll remove it and we can see how it goes in practice.

File: src/compiler/checker.ts (hunk 30 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Giving it a go :smile:

File: src/compiler/checker.ts (hunk 29 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Here's the test we have that I'm removing:
```ts
//// function f(a,b) { }
//// /** @type {f} */
//// var v/**/

goTo.marker();
verify.quickInfoIs('var v: (a: any, b: any) => void');
```

File: src/compiler/checker.ts (hunk 28 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: We actually don't have enough room left in `parserContextFlags`...

File: src/compiler/checker.ts (hunk 30 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: These turn out to not be equivalent for reasons I have not yet investigated

File: src/compiler/checker.ts (hunk 31 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: :heavy_check_mark:

File: src/compiler/checker.ts (hunk 28 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Logged #6578 to address this at a later date
